Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -257,6 +258,83 @@ public async Task<BoolResult> RequestCopyFileAsync(OperationContext context, Con
             }
         }
 
+        /// <summary>
+        /// Pushes content to another machine. Failure to open the source stream should return a null stream.
+        /// </summary>
+        public async Task<BoolResult> PushFileAsync(OperationContext context, ContentHash hash, Func<Task<Stream>> source)
+        {
+            try
+            {
+                var pushRequest = new PushRequest(hash, context.TracingContext.Id);
+                var headers = pushRequest.GetMetadata();
+
+                var call = _client.PushFile(headers, cancellationToken: context.Token);
+                var requestStream = call.RequestStream;
+
+                var responseHeaders = await call.ResponseHeadersAsync;
+
+                var pushResponse = PushResponse.FromMetadata(responseHeaders);
+                if (!pushResponse.ShouldCopy)
+                {
+                    context.TraceDebug($"{nameof(PushFileAsync)}: copy of {hash.ToShortString()} was skipped.");
+                    return BoolResult.Success;
+                }
+
+                var stream = await source();
+
+                if (stream == null)
+                {
+                    await requestStream.CompleteAsync();
+                    return new BoolResult("Failed to retrieve source stream.");
+                }
+
+                using (stream)

```

Comment:
Okay, here is using. This looks right to me, including for errors and cancellations.